panic in interface to interface type assertion

The following program `sample.go` triggers a panic:

```go
package main

import (
	"fmt"
)

type X struct{}

func (X) Foo() int {
	return 1
}
func (X) Bar() int {
	return 2
}

type Foo interface {
	Foo() int
}
type Bar interface {
	Bar() int
}

func main() {
	var x X
	var i Foo = x
	j := i.(Bar)

	fmt.Println(j.Bar())
}
```

Expected result:
```console
$ go run ./sample.go
2
```

Got:
```console
$ yaegi ./sample.go
./sample.go:24:6: panic
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x65dffd]

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc00026c0a0, 0xc00027e500)
	/home/max/go/src/github.com/containous/yaegi/interp/run.go:112 +0x218
panic(0xce6800, 0x15215f0)
	/usr/local/go-1.13.3/src/runtime/panic.go:679 +0x1b2
github.com/containous/yaegi/interp.getMethodByName.func1(0xc00026c0a0, 0xc000278660)
	/home/max/go/src/github.com/containous/yaegi/interp/run.go:1094 +0x14d
github.com/containous/yaegi/interp.runCfg(0xc00027e500, 0xc00026c0a0)
	/home/max/go/src/github.com/containous/yaegi/interp/run.go:118 +0x7a
github.com/containous/yaegi/interp.(*Interpreter).run(0xc000268000, 0xc00027dd00, 0xc00026c000)
	/home/max/go/src/github.com/containous/yaegi/interp/run.go:96 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000268000, 0xc000272000, 0x108, 0x308, 0xc000272000, 0x108, 0x0, 0x1)
	/home/max/go/src/github.com/containous/yaegi/interp/interp.go:360 +0x42d
main.main()
	/home/max/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:82 +0x42d
```

--- a/_test/interface12.go
+++ b/_test/interface12.go
@@ -1,0 +1,32 @@
+package main
+
+import (
+	"fmt"
+)
+
+type X struct{}
+
+func (X) Foo() int {
+	return 1
+}
+func (X) Bar() int {
+	return 2
+}
+
+type Foo interface {
+	Foo() int
+}
+type Bar interface {
+	Bar() int
+}
+
+func main() {
+	var x X
+	var i Foo = x
+	j := i.(Bar)
+
+	fmt.Println(j.Bar())
+}
+
+// Output:
+// 2

--- a/interp/run.go
+++ b/interp/run.go
@@ -120,26 +120,26 @@ func runCfg(n *node, f *frame) {
 }
 
 func typeAssert(n *node) {
-	value := genValue(n.child[0])
-	i := n.findex
+	value := genValue(n.child[0]) // input value
+	dest := genValue(n)           // returned result
 	next := getExec(n.tnext)
 
 	switch {
 	case n.child[0].typ.cat == valueT:
 		n.exec = func(f *frame) bltn {
-			f.data[i].Set(value(f).Elem())
+			dest(f).Set(value(f).Elem())
 			return next
 		}
 	case n.child[1].typ.cat == interfaceT:
 		n.exec = func(f *frame) bltn {
 			v := value(f).Interface().(valueInterface)
-			f.data[i] = reflect.ValueOf(valueInterface{v.node, v.value})
+			dest(f).Set(reflect.ValueOf(valueInterface{v.node, v.value}))
 			return next
 		}
 	default:
 		n.exec = func(f *frame) bltn {
 			v := value(f).Interface().(valueInterface)
-			f.data[i].Set(v.value)
+			dest(f).Set(v.value)
 			return next
 		}
 	}

--- a/interp/value.go
+++ b/interp/value.go
@@ -159,7 +159,18 @@ func genValueInterface(n *node) func(*frame) reflect.Value {
 	value := genValue(n)
 
 	return func(f *frame) reflect.Value {
-		return reflect.ValueOf(valueInterface{n, value(f)})
+		v := value(f)
+		nod := n
+		for {
+			// traverse interface indirections to find out concrete type
+			vi, ok := v.Interface().(valueInterface)
+			if !ok {
+				break
+			}
+			v = vi.value
+			nod = vi.node
+		}
+		return reflect.ValueOf(valueInterface{nod, v})
 	}
 }
 

